onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc

```diff
@@ -123,17 +123,31 @@ void UnaryOpQDQRules(SelectorActionRegistry& qdq_selector_action_registry) {
 void BinaryOpQDQRules(SelectorActionRegistry& qdq_selector_action_registry) {
   // 4 nodes. 2 x DQ for inputs, target, Q
   // Replace with internal QLinear version of operator. Delete all original nodes.
-  const std::string action_name{"2DQ"};
-  std::unique_ptr<Action> action = std::make_unique<QDQ::BinaryReplaceWithQLinear>(kMSDomain);
 
 #if !defined(ORT_MINIMAL_BUILD)
-  // TODO: Enable 16-bit types in selector when binary QLinear* ops support 16-bit.
-  std::unique_ptr<NodeSelector> selector = std::make_unique<QDQ::BinarySelector>();
-  qdq_selector_action_registry.RegisterSelectorAndAction(action_name,
-                                                         {{"Add", {}},
-                                                          {"Mul", {}}},
-                                                         std::move(selector),
-                                                         std::move(action));
+  {
+    const std::string action_name{"2DQ_Mul"};
```

Comment:
changing the name like this could break existing ORT format models, so we should keep the original name.

Reply:
'action_name' is slightly misleading here. It's the combination of selector and action that is being named.

Limiting the existing selector with Mul and Add to the CPU EP (current usage) and adding a new one with just Add that is enabled for the DML EP is one approach but is not great if in the future they can merge back together, as you'd need the new name to still be supported.

An alternative option might be to make the action used here EP aware so that it only modifies the graph for specific EP/s. Could maybe derive from BinaryReplaceWithQLinear to plugin that logic. That way any DML specific logic is included in the build vs the selection side of things where we exclude the selection logic from a minimal build and store the nodes that would have been selected in the ORT format model with that info keyed on the name of the registration. I'd be inclined to do it in a derived class and avoid changing lower level 'action' related code until a case can be made that this is generic infrastructure that will be required by more actions in the future.

Ideally it would be better to do this during GetCapability so the core code doesn't need EP specific logic in it but that would mean including the selection code in a minimal build when the DML EP was enabled as part of that (at least initially). So instead of an L2 transformer you'd call the selector, and if it matched you'd return a ComputeCapability for the fusion to do.

Reply:
Thanks Scott for helping to understand all these options.  Based on this, I believe adding independent Add and Mul registrations for the DML EP alongside existing ones may make the most sense right now.  Quantized operator support tends to be driven by customers and scenarios, and so the set supported by the DML EP may not naturally evolve in lock step, preventing re-merging of registrations as you mentioned. Rather than specializing or refactoring each action class ad hoc, perhaps splitting registrations could actually lead toward filtering them declaratively according to a simple EP interface in the future?

If or when the DML EP deeply supports the ORT model format, though, we may be in a much better place to consider refactoring and understand the implications. For example, the full impact of having stable versus variable behavior of named actions across releases is not currently clear to me.